In [1]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

# Importing Datasets

In [2]:
#Importing Dataset

path= r'/Users/alejandroguillen/Documents/Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_updated.pkl'))

# Creating subset of the first one million entries

In [4]:
df = ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 18)

In [6]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  
0  Regularly busy  Regular days        Average orders  
1  Regularly busy  Slowest days        Average orders  
2  Regularly busy  Slowest days           Most orders  
3      Least busy  Slowest days        Average orders  
4      Least busy  Slowest days           Most orders

In [7]:
df.groupby('product_name')

# Aggregating data with agg

In [8]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
# Multiple aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Aggregating data with transform

In [11]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  max_order  
0  Regularly busy  Regular days        Average orders         10  
1  Regularly busy  Slowest days        Average orders         10  
2  Regularly busy  Slowest days           Most orders         10  
3      Least busy  Slowest days        Average orders         10  
4      Least busy  Slowest days           Most orders         10

In [13]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
..       ...      ...           ...                 ...                ...   
95   3226575      360             1                   5                 12   
96   1469869      377             3                   5                 17   
97   1927023      387             2                   4                 10   
98    858092      420             4                   1                 19   
99    119314      420             5                   3                 19   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
..                    ...         ...                ...        ...   
95                    NaN         196                  1          0   
96                    3.0         196                  9          0   
97                   22.0         196                  3          0   
98                   30.0         196                  2          0   
99                   16.0         196                  2          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
..          ...       ...            ...     ...    ...                ...   
95         Soda        77              7     9.0   both  Mid-range product   
96         Soda        77              7     9.0   both  Mid-range product   
97         Soda        77              7     9.0   both  Mid-range product   
98         Soda        77              7     9.0   both  Mid-range product   
99         Soda        77              7     9.0   both  Mid-range product   

       busiest_day  busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regular days        Average orders         10  
1   Regularly busy  Slowest days        Average orders         10  
2   Regularly busy  Slowest days           Most orders         10  
3       Least busy  Slowest days        Average orders         10  
4       Least busy  Slowest days           Most orders         10  
..             ...           ...                   ...        ...  
95  Regularly busy  Regular days           Most orders          3  
96  Regularly busy  Regular days           Most orders          3  
97      Least busy  Slowest days           Most orders          8  
98  Regularly busy  Busiest days        Average orders         22  
99  Regularly busy  Slowest days        Average orders         22  

[100 rows x 19 columns]

In [14]:
pd.options.display.max_rows = None

In [15]:
#Checking output
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

# Creating flag column for loyalty customers

In [16]:
#Deriving columns with the loc funtion
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [17]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [18]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [20]:
#Checking output
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# 4.8 

## Step 1: done

## Step 2

### In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [21]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3

### Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [22]:
#Previous subset
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

### Observation: with the entire dataframe there is a bigger span in averages (15.2 through 22.9) while with the subset is more limited (with exception of 11.29). The entire dataset allows us to see the shape of the averages more clearly.

# Step 4: Done (check output 11 through 20)

# Step 5:Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer).

In [23]:
# Group by "prices" 
ords_prods_merge.groupby('prices')

In [24]:
# Split the data into groups by "loyalty_flag" and then calculate the mean for each "prices" column
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                      mean
loyalty_flag              
Loyal customer    7.772831
New customer      7.800320
Regular customer  7.797431

### Observation: New Customers have the highest price mean.

# Step 6: Create a spending flag for each user based on the average price across all their orders

In [25]:
# Group data by "user_id" column, and create a mean/product aggregation and column
ords_prods_merge['mean_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [26]:
# Check output
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy  Regular days        Average orders        10  New customer   
1  Regularly busy  Slowest days        Average orders        10  New customer   
2  Regularly busy  Slowest days           Most orders        10  New customer   
3      Least busy  Slowest days        Average orders        10  New customer   
4      Least busy  Slowest days           Most orders        10  New customer   

  mean_price  
0   6.367797  
1   6.367797  
2   6.367797  
3   6.367797  
4   6.367797  

[5 rows x 21 columns]

In [27]:
# Create flags 
ords_prods_merge.loc[ords_prods_merge['mean_price'] < 10, 'spending_flag'] = 'Low spender'

In [28]:
ords_prods_merge.loc[ords_prods_merge['mean_price'] >= 10, 'spending_flag'] = 'High spender'

In [29]:
ords_prods_merge['spending_flag'].value_counts (dropna=False)

Low spender     32285165
High spender      119694
Name: spending_flag, dtype: int64

In [30]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

# Step 7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.

In [31]:
# Group data by "user_id" column
ords_prods_merge['median_on_days'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [32]:
# Check output
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [33]:
# Create flags 
ords_prods_merge.loc[ords_prods_merge['median_on_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [34]:
ords_prods_merge.loc[(ords_prods_merge['median_on_days'] > 10) & (ords_prods_merge['median_on_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [35]:
ords_prods_merge.loc[ords_prods_merge['median_on_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [36]:
# Check output
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: frequency_flag, dtype: int64

In [37]:
# Check null values
print(ords_prods_merge[ords_prods_merge.frequency_flag.isnull()])

          order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_last_order  product_id  \
13645692                 17                    NaN       10749   
17251990                 17                    NaN       33401   
17622767                 17                    NaN       23695   
24138593                 17                    NaN       21334   
25880002                 17                    NaN       22198   

          add_to_cart_order  reordered  \
13645692                  3          0   
17251990                  6          0   
17622767                  2          0   
24138593                  5          0   
25

### Observation: the null values are from customers who have only ordered once.

In [38]:
# Changing the null values to "Non-frequent customer"
ords_prods_merge['frequency_flag'].fillna("Non-frequent customer", inplace = True)

In [39]:
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636442
Name: frequency_flag, dtype: int64

In [40]:
# Export 
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_aggregated.pkl'))